modules/utils/graph/Redif.cpp

```diff
@@ -0,0 +1,325 @@
+			LArray.push_back(graphLap);
+			blaze::DynamicMatrix<Tv> D(n, n);
+			for (size_t i = 0; i < D.rows(); i++)
+				D(i, i) = graphLap(i, i);
+			blaze::DynamicMatrix<Tv> invLeft(n, n);
+			invLeft = blaze::inv(D + 0.25*graphLap);
+			xTrainDe = invLeft * (D*xTrainDe);
+
+			//cout << endl << endl << endl << endl << endl << endl;
+			//cout << i << "-th" << endl;
+			//for (size_t j = 0; j < 10; j++)
+			//	cout << graphLap(0, j) << "\t";
+		}
+
+	}
+
+	template <typename recType, class Metric, typename Tv>
+	blaze::DynamicMatrix<Tv> Redif<recType, Metric, Tv>::getLocalDistMatrix(blaze::DynamicMatrix<Tv> dataSample)
+	{
+		size_t n = dataSample.rows();
+
+		blaze::DynamicMatrix<size_t> ixx(nNeighbors, n);
+		for (size_t i = 0; i < nNeighbors; i++)
+			for (size_t j = 0; j < n; j++)
+				ixx(i, j) = j;
+
+		blaze::DynamicMatrix<Tv> dist(n, n);
+		for (size_t i = 0; i < n; i++)
+			for (size_t j = 0; j < n; j++)
+			{
+				if (i == j)
+				{
+					dist(i, j) = 0;
+					continue;
+				}
+				blaze::DynamicVector<Tv, rowVector> rowi = row(dataSample, i);
+				blaze::DynamicVector<Tv, rowVector> rowj = row(dataSample, j);
+
+
+				std::vector<Tv> veci, vecj;
+				for (size_t k = 0; k < rowi.size(); ++k)
+				{
+					veci.push_back(rowi[k]);
+					vecj.push_back(rowj[k]);
+				}				
+
+				Metric metric;
+				dist(i, j) = metric(veci, vecj);
+			}
+		blaze::DynamicMatrix<size_t> knnMat(nNeighbors, n);
+		blaze::DynamicMatrix<Tv> knnDistMat(nNeighbors, n);
+
+		for (size_t i = 0; i < n; i++)
+		{
+			blaze::DynamicVector<size_t> tempIdx(n);
+			blaze::DynamicVector<Tv> tempDist(n);
+			for (size_t j = 0; j < n; j++)
+			{
+				tempIdx[j] = j;
+				tempDist[j] = dist(j, i);
+			}
+			std::cout << i << "-th" << std::endl;
+			Quicksort(tempIdx, tempDist, 0, n - 1);
+			//if (i == 58)
+			//{
+			//	std::cout << std::endl;
+			//	for (size_t j = 0; j < 30; j++)
+			//		std::cout << tempDist[j] << std::endl;
+			//}
+			for (size_t j = 0; j < nNeighbors; j++)
+			{
+				knnMat(j, i) = tempIdx[j + 1];
+				knnDistMat(j, i) = tempDist[j + 1];
+			}
+		}
+
+		blaze::DynamicMatrix<Tv> localDist(n, n);
+
+		for (size_t i = 0; i < nNeighbors; i++)
+			for (size_t j = 0; j < n; j++)
+				localDist(ixx(i, j), knnMat(i, j)) = knnDistMat(i, j);
+
+		blaze::DynamicMatrix<Tv> retLocalDist(n, n);
+		for (size_t i = 0; i < n; i++)
+			for (size_t j = 0; j < n; j++)
+				retLocalDist(i, j) = (localDist(i, j) > localDist(j, i)) ? localDist(i, j) : localDist(j, i);
+
+		return retLocalDist;
+	}
+
+	template <typename recType, class Metric, typename Tv>
+	blaze::DynamicMatrix<Tv> Redif<recType, Metric, Tv>::calcWeightedGraphLaplacian(blaze::DynamicMatrix<Tv> localDist)
+	{
+		size_t n = localDist.rows();
+
+		/// Degree matrix
+		blaze::DynamicMatrix<Tv> D(n, n);
+		for (size_t i = 0; i < D.rows(); i++)
+		{
+			double sum = 0;
+			for (size_t j = 0; j < n; j++)
+				sum += localDist(i, j);
+			if (sum == 0) sum = 1;
+			D(i, i) = 1 / sum;

```

Comment:
beware of plain numbers without types in combination with templates. They are always integers. make Tv(1) for example. and check, if it really works with integers. implements some checks for incompatible types if_integral .... etc.